{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "239ee4b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def DeBruijn (kmers):\n",
    "    \"\"\"Input : list with str \n",
    "       Output : Dic with str as keys (a kmers' prefix) and LIST that contain str as values ( same kmers' suffix)\n",
    "       \n",
    "       For each kmer, create a key with its prefix (kmer[:-1]) and a value with its suffix (kmer[1:]).\n",
    "       For instance, if AAGA is a kmer, dic[AAG] = AGA. \n",
    "       If a kmer appears multiple times, append its suffix as value for an equal number of times\n",
    "       Note that each key/vertex is a suffix of a kmer and a prefix of another (apart from start & end in RIGHT sequence)\n",
    "       Thus, DeBruijn's output can be used to 'glue' kmers \"\"\"\n",
    "    \n",
    "    adjacency_dic = {}\n",
    "    \n",
    "    for kmer in kmers:\n",
    "        if kmer[:-1] not in adjacency_dic:\n",
    "            adjacency_dic[kmer[:-1]] = []                                    #initiate key with an empty list\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "        kmer_prfx = kmer[:-1]\n",
    "        kmer_sfx  = kmer[1:]\n",
    "        (adjacency_dic[kmer_prfx]).append(kmer_sfx)\n",
    "        \n",
    "    return adjacency_dic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "09a60341",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Contiguous_and_Endpoint_Nodes (adjacency_dic) :\n",
    "    \"\"\" Input : Dic with str (or int) as keys (a kmers' prefix) and LIST with str (or int) as values (same kmers' suffix). \n",
    "        Output: list containing 2 lists as elements, which comprise str or int. First list element corresponds to continue nodes.\n",
    "                For a node to be continue, its has to satisfy this rule : both indegree and outdegree = 1. \n",
    "                Second list element corresponds to nodes that indegree or outdegree != 1 \n",
    "                \n",
    "                In a Debruijn graph, for each of its nodes, assess whether it has more than one input or output.\n",
    "                Separate continue nodes from endpoint nodes.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    outdegreeSum = dict.fromkeys(adjacency_dic, 0)\n",
    "    indegreeSum  = outdegreeSum.copy()\n",
    "    more_indegrees_extra = []\n",
    "    start_or_stop_nodes = []\n",
    "    continue_nodes = []\n",
    "    \n",
    "\n",
    "    for key, values in adjacency_dic.items():                              \n",
    "        outdegreeSum[key] += len(values)                      # add N of outdegrees for each key-node\n",
    "        \n",
    "    for key,values in adjacency_dic.items():                               \n",
    "        for node in values:\n",
    "            try :\n",
    "                indegreeSum[node] += 1                       # remove 1 every time a node is an indegree\n",
    "            except:\n",
    "                more_indegrees_extra.append(node)                       #create artificial node\n",
    "\n",
    "    for key in outdegreeSum:\n",
    "        if outdegreeSum[key]==1 and indegreeSum[key]== 1:\n",
    "            continue_nodes.append(key)\n",
    "        if outdegreeSum[key]!=1 or indegreeSum[key]!= 1 :\n",
    "            start_or_stop_nodes.append(key)\n",
    "            \n",
    "    return [continue_nodes, start_or_stop_nodes + more_indegrees_extra]       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e107c9fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def MaximalNonBranchingPaths (adjacency_dic):\n",
    "    \"\"\"Input 1: Dic with str (or int) as keys (a kmers' prefix) and LIST with str (or int) as values (same kmers' suffix) \n",
    "       Output : List with lists as elements. Every list corresponds to a maximum path without branches at its intermediate nodes\"\"\"\n",
    "    \n",
    "    import copy\n",
    "    \n",
    "    graph = copy.deepcopy(adjacency_dic)\n",
    "    intermediates, endpoints = Contiguous_and_Endpoint_Nodes (adjacency_dic)\n",
    "    All_paths = []\n",
    "    \n",
    "    for start_node in endpoints :\n",
    "        try:\n",
    "            graph[start_node]                                                                                                               \n",
    "        \n",
    "        except KeyError:                                         # some nodes might have only outdegrees,thus appear only as values and not keys\n",
    "            continue                                             ## in this case, jump to next node, since no path can be created by a node which does not constitute a key\n",
    "        \n",
    "        else:   \n",
    "            for i in range(len(graph[start_node])):              #a key/node forms as many paths as the values it projects to\n",
    "                current_node = start_node        \n",
    "                path = [current_node]                            # paths start by one key and continue solely based on values\n",
    "                while True:\n",
    "                    current_node = graph[current_node].pop()     # turn the value to a new key of which a new value will be appended\n",
    "                    path.append(current_node)                    # at least one value after the first node/key will be appended to a path\n",
    "                    if current_node in endpoints:                # stop when there is a node with 0 outputs OR ( more than 1 output or input)\n",
    "                        All_paths.append(path)                   # add the path to the list of paths\n",
    "                        break                                    # break while loop and jump to 2nd for loop if another path exists for the same key \n",
    "                                                                 # Following break, jump to 1st for loop, thus to another key if another path does not exist or the same key. \n",
    "                    else :\n",
    "                        intermediates.remove (current_node)      # Removing the intermediate will facilitate finding isolated cycles later on\n",
    "    \n",
    "    for start_node in intermediates :\n",
    "        current_node = start_node\n",
    "        path = [current_node]\n",
    "        while True:\n",
    "            try :\n",
    "                current_node = graph[current_node].pop()\n",
    "            \n",
    "            except IndexError :                                  # path stops when a key has an empty list (IndexError) as value\n",
    "                if current_node == start_node and len(path) > 1: # if the end value is the same as the start key. Path > 1 to rule out submitting a key itself as a path\n",
    "                    All_paths.append(path) \n",
    "                break\n",
    "            \n",
    "            else : path.append(current_node)\n",
    "\n",
    "    return All_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "db2e005b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PathToGenome(reads) :\n",
    "    \"\"\"Input : list with strings as elements (reads)\n",
    "       Output: string corresponding to the whole genome\n",
    "    \n",
    "    Construct genome based on reads(that is, pieces of genome).\n",
    "    A genome path must have already been constructed to call this function. Hence, all kmers are already found and placed\n",
    "    in the right succession and merely need to be 'joined' into one string.\n",
    "    \n",
    "    Below an explanation of how the input is produced: \n",
    "    \n",
    "       1) Find the first read in the genome and add it as the starting one\n",
    "       2) Find the next read by matching the last already added read (without its first base) \n",
    "          to one of the remaining reads (without their last base)\n",
    "       3) Add the matching read (including its last letter and repeat step 2)\n",
    "               Example of constructed genome based on reads\n",
    "                      ACCGA -> CCGAA ->  CGAAG -> GAAGC\n",
    "                             ||\n",
    "                             \\/\n",
    "                         matched : CCGA                            \"\"\"\n",
    "    \n",
    "    reads_len = len(reads)\n",
    "    \n",
    "    genome = reads[0]\n",
    "    \n",
    "    for read_index in range(1,reads_len):\n",
    "        genome += reads[read_index][-1]\n",
    "    \n",
    "        \n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "187908e6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
